chainer/training/updater.py

```diff
@@ -0,0 +1,366 @@
+    loss_func = loss_func or optimizer.target
+
+    def update(_):
+        batch = iterator.next()
+        in_arrays = converter(batch, device)
+
+        if isinstance(in_arrays, tuple):
+            in_vars = tuple(variable.Variable(x) for x in in_arrays)
+            optimizer.update(loss_func, *in_vars)
+        elif isinstance(in_arrays, dict):
+            in_vars = {key: variable.Variable(x)
+                       for key, x in six.iteritems(in_arrays)}
+            optimizer.update(loss_func, **in_vars)
+        else:
+            in_var = variable.Variable(in_arrays)
+            optimizer.update(loss_func, in_var)
+
+    return update
+
+
+class ParallelUpdater(StandardUpdater):
+
+    """Implementation of a parallel GPU Updater.
+
+    This is an implementation of :class:`Updater` that uses multiple GPUs.
+    It accepts one or more training datasets and one or more optimizers. The
+    default update routine assumes that there is only one training dataset and
+    one optimizer, while users can specify their own update routines. Each
+    batch is converted to input arrays by
+    :func:`~chainer.datasets.concat_examples` by default, which can also be
+    manually set.
+
+    There are two ways to modify the update behavior besides setting a custom
+    loss function. One is by setting a custom update function via the
+    ``update_func`` argument. The other one is by inheriting this class and
+    overriding the :meth:`update` method. In latter case, do not forget to
+    update the iteration counter at each call of this method, because this
+    value is watched by the trainer for deciding when to invoke extensions and
+    when to exit the training loop.
+
+    Args:
+        iterator: Dataset iterator for the training dataset. It can also be a
+            dictionary of iterators. If this is just an iterator, then the
+            iterator is registered by the name ``'main'``.
+        optimizer: Optimizer to update parameters. It can also be a dictionary
+            of optimizers. If this is just an optimizer, then the optimizer is
+            registered by the name ``'main'``.
+        update_func: Update routine. This is a function that takes the updater
+            object as the argument. The default routine uses ``converter``if
+            specified.
+        converter: Converter function to build input arrays. Each batch
+            extracted by the main iterator is split equaly between the
+            devices and then passed with corresponding ``device`` option to
+            this function. If it is omitted,
+            :func:`~chainer.dataset.concat_examples` is used.
+        models: Dictionary of models. The main model should be the same model
+            attached to the ``'main'`` optimizer.
+        devices: Dictionary of devices to which the training data is sent. The
+            devices should be arranged in a dictionary with the same structure
+            as ``models``.
+
```

Comment:
We should note in somewhere that it is "data parallel" scheme that default update routine that `_parallel_update` creates.